:::{.callout-note}

This example code is taken from the fastai [docs](https://docs.fast.ai/tutorial.transformers.html)

:::

In [ ]:
#| include: false
from transformers import GPT2LMHeadModel, GPT2TokenizerFast
from transformers.pytorch_utils import Conv1D
from fastai.text.all import *
import fastcore
from fasterai.sparse.all import *

In [ ]:
pretrained_weights = 'gpt2'
tokenizer = GPT2TokenizerFast.from_pretrained(pretrained_weights)
model = GPT2LMHeadModel.from_pretrained(pretrained_weights)

In [ ]:
path = untar_data(URLs.WIKITEXT_TINY)

In [ ]:
#| include: false
df_train = pd.read_csv(path/'train.csv', header=None)
df_valid = pd.read_csv(path/'test.csv', header=None)

In [ ]:
#| include: false
all_texts = np.concatenate([df_train[0].values, df_valid[0].values])

In [ ]:
#| include: false
class TransformersTokenizer(Transform):
    def __init__(self, tokenizer): self.tokenizer = tokenizer
    def encodes(self, x): 
        toks = self.tokenizer.tokenize(x)
        return tensor(self.tokenizer.convert_tokens_to_ids(toks))
    def decodes(self, x): return TitledStr(self.tokenizer.decode(x.cpu().numpy()))

In [ ]:
#| include: false
splits = [range_of(df_train), list(range(len(df_train), len(all_texts)))]
tls = TfmdLists(all_texts, TransformersTokenizer(tokenizer), splits=splits, dl_type=LMDataLoader)

In [ ]:
#| include: false
bs,sl = 4,256
dls = tls.dataloaders(bs=bs, seq_len=sl)

In [ ]:
#| include: false
def tokenize(text):
    toks = tokenizer.tokenize(text)
    return tensor(tokenizer.convert_tokens_to_ids(toks))

tokenized = [tokenize(t) for t in progress_bar(all_texts)]

In [ ]:
#| include: false
class TransformersTokenizer(Transform):
    def __init__(self, tokenizer): self.tokenizer = tokenizer
    def encodes(self, x): 
        return x if isinstance(x, Tensor) else tokenize(x)
        
    def decodes(self, x): return TitledStr(self.tokenizer.decode(x.cpu().numpy()))

In [ ]:
#| include: false
tls = TfmdLists(tokenized, TransformersTokenizer(tokenizer), splits=splits, dl_type=LMDataLoader)
dls = tls.dataloaders(bs=bs, seq_len=sl)

In [ ]:
#| include: false
class DropOutput(Callback):
    def after_pred(self): self.learn.pred = self.pred[0]

Let's create our fastai `Learner`.

In [ ]:
learn = Learner(dls, model, loss_func=CrossEntropyLossFlat(), cbs=[DropOutput], metrics=Perplexity())

And let's try to extend a given prompt with the pretrained model.

In [ ]:
prompt = "\n = Unicorn = \n \n A unicorn is a magical creature with a rainbow tail and a horn"

In [ ]:
#| include: false
prompt_ids = tokenizer.encode(prompt)
inp = tensor(prompt_ids)[None]

In [ ]:
preds = learn.model.generate(inp, max_length=40, num_beams=5, temperature=1.5)

In [ ]:
tokenizer.decode(preds[0].cpu().numpy())

'\n = Unicorn = \n \n A unicorn is a magical creature with a rainbow tail and a horn on its head.\n\nA unicorn is a magical creature with a rainbow tail and a horn'

In [ ]:
learn.validate()

[3.695716142654419, 40.27440643310547]

In [ ]:
learn.fit_one_cycle(1, 1e-4)

epoch     train_loss  valid_loss  perplexity  time
0         3.155010    2.842319    17.155495   01:12


In [ ]:
prompt_ids = tokenizer.encode(prompt)
inp = tensor(prompt_ids)[None]

preds = learn.model.generate(inp.cuda(), max_length=40, num_beams=5, temperature=1.5)

tokenizer.decode(preds[0].cpu().numpy())

'\n = Unicorn = \n \n A unicorn is a magical creature with a rainbow tail and a horn @-@ like head . It is the most common unicorn in the world . Unicorn horns are'

## Make it sparse !

Let's see now if we retrain our model, this time introducing sparsity

In [ ]:
learn = Learner(dls, model, loss_func=CrossEntropyLossFlat(), cbs=[DropOutput], metrics=Perplexity())

Unfortunately, the transformer model uses a custom layer: `Conv1D`, which is not a part of PyTorch. To overcome this problem, we have to add this layer to our `Granularities` class, so that it knows what to sparsify.

Here, the `Conv1D` behaves like a `Linear` layer: its weight is a matrix of shape `(nx, nf)`, input features by output features.

In [ ]:
doc(Conv1D)

We can thus add the Conv1D granularity by using the `add_granularity` method, indicating the target module and the corresponding granularities that it can handle (the same as `Linear` so we can reuse it)

In [ ]:
Granularities.add_granularity(Conv1D, Granularities._granularities_Linear)

Let's now define our `SparsifyCallback`. We make the model 30% sparse, weight by weight, each `Conv1D` layer treated independently (`context='local'`), zeroing the lowest-magnitude weights — this is what `large_final` selects, as it keeps the largest ones.

In [ ]:
sp_cb = SparsifyCallback(sparsity=30, granularity='weight', context='local', criteria=large_final, schedule=one_cycle, layer_type=Conv1D)

We now only have to pass our callback to fastai

In [ ]:
learn.fit_one_cycle(1, 1e-4, cbs=sp_cb)

Sparsifying weight until a sparsity of 30%
epoch     train_loss  valid_loss  perplexity  time
Saving Weights at epoch 0
Sparsity at the end of epoch 0: 30.00%
0         3.088130    2.861551    17.488626   05:30
Final Sparsity: 30.00%
Sparsity Report:
--------------------------------------------------------------------------------
Layer                          Type            Params     Zeros      Sparsity
--------------------------------------------------------------------------------
transformer.h.0.attn.c_attn    Conv1D          1,769,472  530,840       30.00%
transformer.h.0.attn.c_proj    Conv1D          589,824    176,946       30.00%
transformer.h.0.mlp.c_fc       Conv1D          2,359,296  707,787       30.00%
transformer.h.0.mlp.c_proj     Conv1D          2,359,296  707,787       30.00%
transformer.h.1.attn.c_attn    Conv1D          1,769,472  530,840       30.00%
transformer.h.1.attn.c_proj    Conv1D          589,824    176,946       30.00%
transformer.h.1.mlp.c_fc       Conv

And we can check the predicion to the same prompt as before

In [ ]:
prompt_ids = tokenizer.encode(prompt)
inp = tensor(prompt_ids)[None]

preds = learn.model.generate(inp.cuda(), max_length=40, num_beams=5, temperature=1.5)

tokenizer.decode(preds[0].cpu().numpy())

'\n = Unicorn = \n \n A unicorn is a magical creature with a rainbow tail and a horn on its head . The unicorn is a member of the <unk> family of unicorns ,'

That's it! You now have a sparse Transformer. On the wikitext-tiny test split, the pretrained GPT-2 scores a perplexity of 40.27; one epoch of fine-tuning brings it to 17.16, and one epoch of fine-tuning while sparsifying 30% of every `Conv1D` weight gives 17.49. Single run each, no seed spread. The model is neither smaller nor faster as it stands: the zeroed weights are still stored and still multiplied. To turn sparsity into a speed-up, see the [granularity](../../core/granularity.html) page.

---

## Summary

| Tool | Purpose |
|------|---------|
| `Granularities.add_granularity` | Register a custom layer type (`Conv1D`) so fasterai knows how to sparsify it |
| `SparsifyCallback(..., layer_type=Conv1D)` | Sparsify the registered layers during training |
| `large_final` | Criteria keeping the weights with the largest magnitude |
| `one_cycle` | Schedule used here for the sparsity ramp |

## See Also

- [Sparsifier](../../sparse/sparsifier.html) - Core sparsification API
- [SparsifyCallback](../../sparse/sparsify_callback.html) - Callback for training integration
- [Granularity](../../core/granularity.html) - What gets removed, and which granularities can translate to a speed-up
- [Schedules](../../core/schedules.html) - All available sparsification schedules
- [Sparsifier Tutorial](sparsifier.html) - Basic sparsification walkthrough